docs/src/main/asciidoc/hibernate-search-guide.adoc

```diff
@@ -0,0 +1,627 @@
+
+== Adding full text capabilities to our REST service
+
+In our existing `LibraryResource`, we just need to inject the following methods (and add a few ``import``s):
+
+[source,java]
+----
+    @Transactional // <1>
+    void onStart(@Observes StartupEvent ev) throws InterruptedException { // <2>
+        // only reindex if we imported some content
+        if (Book.count() > 0) {
+            Search.getSearchSession(em)
+                    .createIndexer()
+                    .startAndWait();
+        }
+    }
+
+    @GET
+    @Path("author/search")
+    @Transactional
+    public List<Author> searchAuthors(@QueryParam("pattern") String pattern) { // <3>
+        List<Author> authors = Search.getSearchSession(em) // <4>
+                .search(Author.class) // <5>
+                .predicate(f ->
+                    pattern == null || pattern.trim().isEmpty() ?
+                            f.matchAll() : // <6>
+                            f.simpleQueryString()
+                                .onFields("firstName", "lastName", "books.title").matching(pattern) // <7>
+                )
+                .sort(f -> f.byField("lastName_sort").then().byField("firstName_sort")) // <8>
+                .fetchHits(); // <9>
+
+        return authors;
+    }
+----
+<1> Important point: we need a transactional context for these methods.
+<2> As we will import data into the PostgreSQL database using an SQL script, we need to reindex the data at startup.
+For this, we use Hibernate Search's mass indexer, which allows to index a lot of data efficiently (you can fine tune it for better performances).
+All the upcoming updates coming through Hibernate ORM operations will be synchronized automatically to the full text index.
+If you don't import data manually in the database, you don't need that:
+the mass indexer should then only be used when you change your indexing configuration (adding a new field, changing an analyzer's configuration...) and you want the new configuration to be applied to your existing entities.
+<3> This is where the magic begins: just adding the annotations to our entities make them available for full text search: we can now query the index using the Hibernate Search DSL.
+<4> First, we get an Hibernate Search session from the injected entity manager.
+<5> We indicate that we are searching for ``Author``s.
+<6> We create a predicate: if the pattern is empty, we use a `matchAll()` predicate.
+<7> If we have a valid pattern, we create a https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-simple-query-string-query.html[`simpleQueryString()`] predicate on the `firstName`, `lastName` and `books.title` fields matching our pattern.
+<8> We define the sort order of our results. Here we sort by last name, then by first name. Note that we use the specific fields we created for sorting.
+<9> This is self explanatory. Obviously, paging is also supported.
```

Comment:
You might want to use paging here, then. Using `fetchHits` without paging is arguably not something you will use in real-world applications.

Reply:
Well, my UI does not have paging so not really. I don't think it's an issue, paging is easy to use and is not really specific to Hibernate Search.